synapse/util/metrics.py

```diff
@@ -82,16 +89,16 @@ async def foo(...):
 
     """
 
-    def wrapper(func: T) -> T:
+    def wrapper(func: F) -> F:
```

Comment:
Why this change? We lose a bunch of typing information here, and isn't what the mypy docs recommend for annotating decorators.

Reply:
As in [this](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html#decorators)? Hadn't seen this before.

And I don't think I'd properly grokked the difference between a bound typevar and just using the bound verbatim.

----
Motivation: I wanted to encode in the type system that the `func` being decorated was asynchronous, and so too was the decorated version `measured_func`. Maybe there's a better way? Or maybe it's not worth it?

Reply:
I wanted to emphasise that the function being decorated should be asynchronous. But I'm happy to revert this, I don't think is a particularly good job